README.md
# KA-Rando

Kirby's Adventure Randomizer

Randomize levels, character colors, and enemy abilities for *Kirby's Adventure* on NES.

https://austinbricker.com/KA-Rando

https://github.com/aquova/KA-Rando

Written by Austin Bricker, 2017-2019

## -- Overview --

Randomizes the level order, enemy abilities, and Kirby's color in Kirby's Adventure. The main functionality is a web interface which is meant to be used with a US version of Kirby's Adventure for NES. As of Version 3.0.3, only the original release of the game (PRG0) is supported.

## -- Features --

- Randomize the levels! You can now randomize the level order, so each level door you go through will take you to an unexpected level!
    - Note, boss doors and a handful of the regular levels (the ones with warp stars) are not randomized as of yet.
- Randomize enemy abilities! Enemies will give you different abilities than the vanilla game when eaten.
- Change your color! Choose between a few presets I've designed, or pick any color that the NES has to offer!

## -- Usage --

All you need is a US copy of the original version of the Kirby's Adventure ROM (which is left to the user to obtain). Then, visit https://austinbricker.com/KA-Rando, choose the options you desire, select your file, and press 'Randomize'.

## -- Running Locally --

This project runs entirely in the client, so if you need to use the randomizer off-line, you can simply download and visit the webpage in a browser. To do so, either click `Clone or download` on the GitHub page and download the zip file, or in a Terminal run `git clone https://github.com/aquova/KA-Rando`. Once downloaded, simply open `index.html` in a browser of your choice!

## -- Credits --

The source code for the CryptoJS MD5 library can be found here: https://code.google.com/archive/p/crypto-js/#MD5

## -- Known Issues --

- Dying with Star Rod disables select button until level completion
- All six levels with in-level Warp Stars are not randomized, due to game-breaking bugs occurring.
- Museum enemies are not randomized

## -- Version History --

v4.0.0 - 11/1/18 - Ported entire project from Python to JavaScript, from now on the project will use a web-based interface.

v3.0.3 - 2/25/18 - Changed icon to not match KNDL-Randomizer, removed support for PRG1

v3.0.2 - 1/23/18 - Removed the last two stages with Warp Stars from randomization to avoid potential bugs

v3.0.1 - 1/22/18 - Removed stage 7-2 from randomization list due to Warp Star bugs

v3.0.0 - 12/31/17 - Added support for randomizing the level order, full randomization of Kirby's color, and checksum verification

v2.0.1 - 11/20/17 - Added option to only change Kirby's color, but don't randomize anything

v2.0.0 - 11/16/17 - Added GUI interface. Can now select specific Kirby color in addition to randomizing the color. Switched from Python 2 to Python 3

v1.0.3 - 9/2/17 - Removed need to import data from txt files. Entire program now contained in .py

v1.0.2 - 6/8/17 - Added option to randomize Kirby's color palette

v1.0.1 - 5/25/17 - Fixed issues with OS dependancy. The Python program now works on all OS's, and added an exe program for Windows users.

v1.0.0 - 5/24/17 - Completed the Python program. Enemy abilites are now randomized, supports options for omitting the Star Rod ability or randomizing non-ability enemies.


release-notes.md
# Release Notes

## 4.0.2

Fixed issue with the checksum verification.

## 4.0.0

A very big release this time. I'd been getting increasingly frustratated with the GUI setup I had, and decided to bite the bullet and convert to a web interface. I rewrote all the logic and UI in JavaScript, and implemented some of the additions from the sister *Kirby Color Editor* project. In addition to being a web release, all the original functionality is there, as well as a greater ability to make palette swaps.

## 3.0.3

Two small changes to this release

- With the release of my Kirby: Nightmare in Dream Land randomizer, I've decided to use the old icon for that, and a new one from Kirby's Adventure for this one.
- I've decided to limit the randomizer to work only with the original release (PRG0) ROM, and not the first revision (PRG1). This second ROM always had issues with the color palettes, and it's not worth the effort to try and debug two different versions of the game. If you still wish to use this ROM, the older releases still support it, but with their existing bugs.

## 3.0.2

Due to further issues with levels containing Warp Stars, I've removed the final two levels from the randomized list until I can figure out a way to stop the problem

## 3.0.1

A small patch which removes Stage 7-2 from being randomized due to issues with the Warp Star.

## 3.0.0

Another big new release!

- You can now randomize the level order of the game! Entering doors from the level overworlds now lead to completely new random levels. Note that once a level is started, all the rooms are still in the same order as normal.
- You can now also completely randomize Kirby's color! The randomizer will choose random values to use as Kirby's color palette. To be honest, it's really hideous, but it was easy to add, so I figured why not.

Some other smaller things:

- The randomizer now verifies the checksum of the given ROM. If the ROM is not a US Kirby's Adventure initial version or first revision (PRG0 or PRG1) then it will be rejected.
- Currently three levels are not randomized in the level order. They contain in-level Warp Stars which had some game breaking bugs when they were randomized. There are three other levels which do contain Warp Stars, but they seemed to still work as intended. If they prove to also have issues, then they may be removed from the level randomization as well. The boss stages are also no randomized as of yet.
- The CLI version of the program has also been updated to Version 3.0.0, but does not allow to specifically choose Kirby's color.

## 2.0.1

Added new option in the GUI to only change Kirby's color, but don't randomize any behavior.

## 2.0.0

A big new release!

- I've added a GUI interface! Running KA-Randomize now opens a graphical application to select your options.
- In addition to the PC .exe binary, I've also included a .app file for Mac users.
- With the GUI, you can not only randomize Kirby's color, but also have the option of choosing which color you want.
- I've (finally) made the switch from Python 2.7 to Python 3
- For Linux users/those who don't want to use the GUI, a Python 3 version of the original command line interface is also included, in KA-Rando-Simple.py

## 1.0.3

Improved program structure, no longer relies on importing data from external files.

## 1.0.2

Adds option to randomize Kirby's color based on a few preset values.

## 1.0.1

Added Windows support

## 1.0.0

First release of the Kirby's Adventure Randomizer


js/colors.js
// The original KA colors, border, feet, then body
var originalPalette = [
    [3, 1, 3],
    [252, 194, 228],
    [252, 110, 204]
]

// Several new palettes for Kirby, chosen by me
var presetPalettes = [
    ["0D", "35", "25"], // Pink
    ["0D", "25", "15"], // Red
    ["0D", "27", "17"], // Orange
    ["0D", "38", "28"], // Yellow
    ["0D", "39", "29"], // Light Green
    ["0D", "29", "19"], // Green
    ["0D", "31", "21"], // Light Blue
    ["0D", "21", "11"], // Blue
    ["0D", "24", "14"], // Purple
    ["0D", "20", "10"]  // Grayscale
]

// Places to replace Kirby's palette
var colorLocations = [0x32FEC, 0x43AC9, 0x43AF7, 0x43CE1, 0x47E4A, 0x6D5EB, 0x6D5EF, 0x711E4, 0x711EB, 0x76759, 0x76783, 0x7679B, 0x767B3, 0x767CB, 0x767E3, 0x767FB, 0x76813, 0x7682B, 0x7684F, 0x7687B, 0x77A61, 0x78A0D, 0x78A1D, 0x78A2D, 0x78A35, 0x79552, 0x79562, 0x7ADA6, 0x7ADB6, 0x7ADD2, 0x2CED0, 0x2D2F6, 0x43B79, 0x43CE9, 0x5C980, 0x5C98C, 0x69D29, 0x69D49, 0x69D69, 0x69D71, 0x69D89, 0x69DA9, 0x69DC9, 0x69DE9, 0x69E09, 0x6D5F7, 0x6DBAF, 0x6DF56, 0x711F3, 0x79542, 0x7954A, 0x7855A, 0x7ADCA, 0x43ACD]

// Map of NES palette values to 24-bit hex color
var nesMap = {
    "00": "7C7C7C",
    "01": "0000FC",
    "02": "0000BC",
    "03": "4428BC",
    "04": "940084",
    "05": "A80020",
    "06": "A81000",
    "07": "881400",
    "08": "503000",
    "09": "007800",
    "0A": "006800",
    "0B": "005800",
    "0C": "004058",
    "0D": "000000",
    "10": "BCBCBC",
    "11": "0078F8",
    "12": "0058F8",
    "13": "6844FC",
    "14": "D800CC",
    "15": "E40058",
    "16": "F83800",
    "17": "E45C10",
    "18": "AC7C00",
    "19": "00B800",
    "1A": "00A800",
    "1B": "00A844",
    "1C": "008888",
    "1D": "000000",
    "20": "F8F8F8",
    "21": "3CBCFC",
    "22": "6888FC",
    "23": "9878F8",
    "24": "F878F8",
    "25": "F85898",
    "26": "F87858",
    "27": "FCA044",
    "28": "F8B800",
    "29": "B8F818",
    "2A": "58D854",
    "2B": "58F898",
    "2C": "00E8D8",
    "2D": "787878",
    "30": "FCFCFC",
    "31": "A4E4FC",
    "32": "B8B8F8",
    "33": "D8B8F8",
    "34": "F8B8F8",
    "35": "F8A4C0",
    "36": "F0D0B0",
    "37": "FCE0A8",
    "38": "F8D878",
    "3D": "F8D8F8",
    "39": "D8F878",
    "3A": "B8F8B8",
    "3B": "B8F8D8",
    "3C": "00FCFC",
    "3D": "F7D8F8"
}

var nesColors = ["00", "01", "02", "03", "04", "05", "06", "07", "08", "09", "0A", "0B", "0C", "0D", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "1A", "1B", "1C", "1D", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "2A", "2B", "2C", "2D", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "3A", "3B", "3C", "3D"]

var brightText = ["01", "02", "03", "04", "05", "06", "07", "08", "09", "0A", "0B", "0C", "0D", "11", "12", "13", "18", "19", "1A", "1B", "1C", "1D", "2D"]

// There is probably a better way to pass the NES value to be written into the ROM, but I couldn't think of it
var nesVals = ["0F", "35", "25"]

function populateTable() {
    var t = document.getElementById("paletteTable")
    // Table shall be 4 rows, 14 columns
    for (var row = 0; row < 4; row++) {
        var tableRow = document.createElement("tr")
        for (var col = 0; col < 14; col++) {
            var cell = document.createElement("input")
            cell.type = "button"
            cell.id = nesColors[14 * row + col]
            cell.style = "width: 20px"
            cell.style = "background-color: #" + nesMap[nesColors[14 * row + col]]
            cell.addEventListener("click", function(d) {
                var thisButton = document.getElementById("palette").getElementsByTagName("button")[t.value]
                thisButton.style.backgroundColor = d.target.style.backgroundColor
                if (brightText.includes(d.target.id)) {
                    thisButton.style.color = "#fff"
                } else {
                    thisButton.style.color = "#000"
                }
                changeColor(t.value, nesMap[d.target.id])
                nesVals[t.value] = d.target.id
                t.classList.add("hidden")
            })
            tableRow.appendChild(cell)
        }
        t.appendChild(tableRow)
    }
}

function viewTable(index) {
    var t = document.getElementById("paletteTable")

    if (t.classList.contains("hidden")) {
        t.classList.remove("hidden")
    } else if (t.value == index) {
        t.classList.add("hidden")
    }
    t.value = index
}

// Rewrite ROM with new palette
function replaceColor() {
    // Yeah, these are out of order, but I didn't want to redo all of the for loops just for this
    var newColor0 = parseInt(nesVals[1], 16)
    var newColor1 = parseInt(nesVals[2], 16)
    var newColor2 = parseInt(nesVals[0], 16)

    for (var i = 0; i < colorLocations.length; i++) {
        var colorAddress = colorLocations[i]
        rom[colorAddress] = newColor0
        rom[colorAddress + 1] = newColor1
        rom[colorAddress + 2] = newColor2
    }
}

// Draws Kirby on canvas
function drawKirby() {
    ctx.clearRect(0, 0, canvas.width, canvas.height)
    ctx.drawImage(img, 0, 0, img.width, img.height)
    imageData = ctx.getImageData(0, 0, img.width, img.height)
    ctx.drawImage(canvas, 0, 0, img.width, img.height, 0, 0, canvas.width, canvas.height)

    originalPixelArray = Uint8ClampedArray.from(imageData.data)
}

function hex2rgb(hex) {
    var r = parseInt(hex.slice(0, 2), 16)
    var g = parseInt(hex.slice(2, 4), 16)
    var b = parseInt(hex.slice(4, 6), 16)

    return [r, g, b]
}

// Draw recolored Kirby on canvas
// Iterate through pixels of Kirby image, recolor if a match
function changeColor(oldIndex, color) {
    var newColor = hex2rgb(color)

    // Pixel array is four parts: R, G, B, A
    var length = originalPixelArray.length / 4;
    var newPixelArray = Uint8ClampedArray.from(imageData.data)
    for (var i = 0; i < length; i++) {
        var index = 4 * i;

        var r = originalPixelArray[index];
        var g = originalPixelArray[index + 1];
        var b = originalPixelArray[index + 2];

        if (r == originalPalette[oldIndex][0] && g == originalPalette[oldIndex][1] && b == originalPalette[oldIndex][2]) {
            newPixelArray[index] = newColor[0];
            newPixelArray[index + 1] = newColor[1];
            newPixelArray[index + 2] = newColor[2];
        }
    }
    imageData.data.set(newPixelArray)
    ctx.putImageData(imageData, 0, 0);
    ctx.drawImage(canvas, 0, 0, img.width, img.height, 0, 0, canvas.width, canvas.height);
}

function setPreset() {
    var presetIdx = Number(document.getElementById("presets").value)

    var nodes = document.getElementById("palette").getElementsByTagName("button")
    for (var i = 0; i < nodes.length; i++) {
        nesVals[i] = presetPalettes[presetIdx][i]
        nodes[i].style.backgroundColor = "#" + nesMap[presetPalettes[presetIdx][i]]
        if (brightText.includes(presetPalettes[presetIdx][i])) {
            nodes[i].style.color = "#fff"
        } else {
            nodes[i].style.color = "#000"
        }
        changeColor(i, nesMap[presetPalettes[presetIdx][i]])
    }
}
